File: Apps/Desktop/Sources/Terminal/TerminalTabNaming.swift
enum TerminalTabNaming {
    static func title(for number: Int) -> String {
        "Tab \(number)"
    }

    static func nextNumber(existingTitles: some Sequence<String>) -> Int {
        let usedNumbers = Set(existingTitles.compactMap(number(in:)))
        var candidate = 1
        while usedNumbers.contains(candidate) {
            candidate += 1
        }
        return candidate
    }

    static func number(in title: String) -> Int? {
        let prefix = "Tab "
        guard title.hasPrefix(prefix) else { return nil }
        let suffix = String(title.dropFirst(prefix.count))
        guard let number = Int(suffix), number > 0, String(number) == suffix else {
            return nil
        }
        return number
    }
}

File: packages/electron/src/shared/terminalTabNaming.ts
export function title(forNumber: number): string {
  return `Tab ${forNumber}`;
}

export function nextNumber(existingTitles: Iterable<string>): number {
  const used = new Set<number>();
  for (const existing of existingTitles) {
    const parsed = numberIn(existing);
    if (parsed != null) {
      used.add(parsed);
    }
  }
  let candidate = 1;
  while (used.has(candidate)) {
    candidate += 1;
  }
  return candidate;
}

export function numberIn(titleText: string): number | null {
  const prefix = "Tab ";
  if (!titleText.startsWith(prefix)) {
    return null;
  }
  const suffix = titleText.slice(prefix.length);
  const parsed = Number.parseInt(suffix, 10);
  if (!Number.isInteger(parsed) || parsed <= 0 || String(parsed) !== suffix) {
    return null;
  }
  return parsed;
}
